framework/src/modules/interoperability/mainchain/module.ts

```diff
@@ -419,4 +438,108 @@ export class MainchainInteroperabilityModule extends BaseInteroperabilityModule
+		const sortedByChainID = [...terminatedStateAccounts].sort((a, b) =>
+			a.chainID.compare(b.chainID),
+		);
+		for (let i = 0; i < terminatedStateAccounts.length; i += 1) {
+			if (!terminatedStateAccounts[i].chainID.equals(sortedByChainID[i].chainID)) {
+				throw new Error('terminatedStateAccounts must be ordered lexicographically by chainID.');
+			}
+		}
+
+		// For each entry stateAccount in terminatedStateAccounts holds
+		// stateAccount.stateRoot == chainData.lastCertificate.stateRoot,
+		// stateAccount.mainchainStateRoot == EMPTY_HASH, and
+		// stateAccount.initialized == True.
+		// Here chainData is the corresponding entry (i.e., with chainID == stateAccount.chainID) in chainInfos.
+		for (const chainInfo of chainInfos) {
+			const stateAccount = terminatedStateAccounts.find(a =>
+				a.chainID.equals(chainInfo.chainID),
+			)?.terminatedStateAccount;
+			if (stateAccount) {
+				if (!stateAccount.stateRoot.equals(chainInfo.chainData.lastCertificate.stateRoot)) {
+					throw new Error(
+						"stateAccount.stateRoot doesn't match chainInfo.chainData.lastCertificate.stateRoot.",
+					);
+				}
+
+				if (!stateAccount.mainchainStateRoot.equals(EMPTY_HASH)) {
+					throw new Error('stateAccount.mainchainStateRoot is not equal to EMPTY_HASH.');
+				}
+
+				if (!stateAccount.initialized) {
+					throw new Error('stateAccount is not initialized.');
+				}
+			}
+		}
+	}
+
+	// https://github.com/LiskHQ/lips/blob/main/proposals/lip-0045.md#mainchain
+	private _verifyTerminatedOutboxAccounts(
+		_chainInfos: ChainInfo[],
+		terminatedStateAccounts: TerminatedStateAccountWithChainID[],
+		terminatedOutboxAccounts: TerminatedOutboxAccountWithChainID[],
+	) {
+		// Each entry outboxAccount in terminatedOutboxAccounts has a unique outboxAccount.chainID
+		const chainIDs = terminatedOutboxAccounts.map(a => a.chainID);
+		if (!bufferArrayUniqueItems(chainIDs)) {
+			throw new Error(`terminatedOutboxAccounts don't hold unique chainID.`);
+		}
+
+		// terminatedOutboxAccounts is ordered lexicographically by outboxAccount.chainID
+		const sortedByChainID = [...terminatedOutboxAccounts].sort((a, b) =>
+			a.chainID.compare(b.chainID),
+		);
+		for (let i = 0; i < terminatedOutboxAccounts.length; i += 1) {
+			if (!terminatedOutboxAccounts[i].chainID.equals(sortedByChainID[i].chainID)) {
+				throw new Error('terminatedOutboxAccounts must be ordered lexicographically by chainID.');
+			}
+		}
+
+		// Furthermore, an entry outboxAccount in terminatedOutboxAccounts must have a corresponding entry
+		// (i.e., with chainID == outboxAccount.chainID) in terminatedStateAccounts
+		for (const outboxAccount of terminatedOutboxAccounts) {
+			if (
+				terminatedStateAccounts.find(a => a.chainID.equals(outboxAccount.chainID)) === undefined
+			) {
+				let msg =
+					'Each entry outboxAccount in terminatedOutboxAccounts must have a corresponding entry ';
+				msg += '(with chainID == outboxAccount.chainID) in terminatedStateAccounts.';
```

Comment:
not sure why we have this concatenation but maybe we could specifically mention in the error 
```ts
throw new Error(`Each entry outboxAccount in terminatedOutboxAccounts must have a corresponding entry in terminatedStateAccount. outboxAccount with chainID: ${outboxAccount.chainID.toString('hex')} does not exist in terminatedStateAccounts`)
```